### PR TITLE
Feature: Add access resource using the KBS-provisioned Token

### LINKED ISSUE #129: Feature | Access resource with the KBS-provisioned token
Now after the handshake of the RCAR protocol, KBS will response a token to the requester. This token should have access to the KBS' resources. (This avoids to re-establish a RCAR session again after the former session breaks in the timeout of the token)

Now we are using the same session where RCAR handshake happens to access resource. This issue is to normalize this behavior. That means the resource access should be spilit into two:
1. With RCAR protocol, the client get a short-liveness token (currently ~10min)
2. the client then access resource with this token, or with the same session that RCAR uses. KBS should support both of the ways.

cc @sameo @jialez0 Please check the rationale of this, thanks! And I'd like to push this both kbs and AA side.

### PR DESCRIPTION
**Pull request by Xynnn007**
This PR mainly aims to add a new feature that a requester can bring the KBS-provisioned token to access the resources. The rationale behind this is the token is always short-lived, which could help decrease the risk of token leakage.

This PR also refactor the error handling of the `http` module a little to use the question mark `?` for better error handling.

Close #129

### CODE REVIEW

**Comment by Xynnn007** [comment]
About the error handling, there are still a lot of `anyhow` crate used in KBS. KBS protocol document said that we should return different HTTP status code due to different error. I did not dive into this, but let me put an issue to record this.

File: docs/kbs.yaml
Comment (jialez0) [inline docs/kbs.yaml:126]: I think we'd better put the token in the `bear_auth` header of HTTP Req. (Same as what we do in the `set-resource` API)

Comment (Xynnn007) [reply docs/kbs.yaml:126]: Good point. Fixed.

File: src/api/src/http/resource.rs
Comment (jialez0) [inline src/api/src/http/resource.rs:120]: We'd better verify token here locally instead of call `token_broker` to do it, because the issuer may not be this KBS itself. I think we  can set a "get_token_verification_pubkey" API, it can help to get pubkey from  token broker or remote token issuer, which can be specified by config parameter.

Comment (Xynnn007) [reply src/api/src/http/resource.rs:120]: Currently we only suppose this token is provisioned by the same KBS just then, for the token is short-lived ~10min. Do you mean a long-lived token?

Comment (jialez0) [reply src/api/src/http/resource.rs:120]: This is not related to the lifecycle of the token. According to my understanding, in the confidential data hub design, the issuer of the token will not be the same entity as the consumer of the token most of the time, even if it is a token with a short lifecycle, it does not matter.

Comment (Xynnn007) [reply src/api/src/http/resource.rs:120]: Do you mean a scenario like this:
1. Client gets a token provisioned by an attestation service rather than CoCo. For example Amber
2. Client wants to use this token to request resource in KBS

Comment (jialez0) [reply src/api/src/http/resource.rs:120]: > Do you mean a scenario like this:
> 
> 1. Client gets a token provisioned by an attestation service rather than CoCo. For example Amber
> 2. Client wants to use this token to request resource in KBS

Exactly. The token issuer in 1 may not only be Amber, but also CoCo AS deployed by trusted parties (specifically used to issue tokens).

Comment (Xynnn007) [reply src/api/src/http/resource.rs:120]: Ok. I think this should be carefully handled. I think there are two different tokens in the context:
I. In KBS: Signed(claims, pubkey)
II. In Amber: Signed(claims)

We should firstly ensure this token is with the proper token (type I), with which we can encrypt the dataflow of the confidential resource. If the token is type II, we should think about how we send back the confidential resource.

Also, the basic logic of a certificated token is:
1. Verify the endorsement and claims due to the user defined policy
3. A token is provisioned

Also, if tokens provisioned by other CoCo-KBS is to be supported, we should recheck the claims via giving it to the backend CoCo-AS against the policy of the current CoCo-KBS/AS. 

My idea is that we can start with a simple scenario, s.t.
1. This token is provisioned by the same KBS.

To accept a token from another KBS, we should
1. impl new API for AS to receive different signed claims rather than a raw evidence to verify against a policy.
2. Let KBS leverage this new API to forward the signed claims of the token (or directly the token)

To accept a token from another attestation service like Amber, we should at least
1. Let Amber to support a token with pubkey
Or 2. Think of a way where a verified token can be derived into a tee key pair to fit in the KBS protocol to encrypt the dataflow
Or 3. Do some modifications upon the current KBS protocol

Comment (jialez0) [reply src/api/src/http/resource.rs:120]: OK, I agree start with the simple scenario that token provisioned by the same KBS. 

However, two points need to be emphasized:

1. Verify that the claims in the token should not be in AS, but should introduce a new policy engine in KBS, which is consistent with the RATS passport model. The policy here is the policy of rely party verifying the attachment results.
2. After this PR is merged, I will transfer the logic of the Token broker from KBS to AS. In other words, AS no longer returns a results claims, but directly returns the token. Therefore, I may modify the token's verification logic here to: obtain the token verification public key from AS's new API and then have KBS verify the token.

File: docs/kbs_attestation_protocol.md
Comment (sameo) [inline docs/kbs_attestation_protocol.md:342]: I would add the following paragraph to introduce the two authentication scheme:

```
Attesters can access protected resources by providing the KBS with either the HTTP cookie or the
Attestation Results token returned as part of the Attestation HTTP response. The former is usually
short-lived and is typically used for one-time, synchronous resources request. For longer KBS
sessions, it is more convenient to use the latter.    
```

File: docs/kbs_attestation_protocol.md
Comment (sameo) [inline docs/kbs_attestation_protocol.md:344]:
```
### HTTP Cookie Authentication
```

File: docs/kbs_attestation_protocol.md
Comment (sameo) [inline docs/kbs_attestation_protocol.md:389]:
```
### Attestation Results Token Authentication
```

File: docs/kbs_attestation_protocol.md
Comment (sameo) [inline docs/kbs_attestation_protocol.md:394]:
```markdown

As part of the Authentication process, the requester generates a TEE asymmetric key 
pair. Upon successful Attestation, the KBS provisions a token that contains the
Attestation Results and endorsements for the TEE public key.
```

File: docs/kbs_attestation_protocol.md
Comment (sameo) [inline docs/kbs_attestation_protocol.md:399]: Isn't the KBS response encrypted with a symetric key, that is wrapped with the attestation results TEE key?

File: src/api/Cargo.toml
Comment (sameo) [inline src/api/Cargo.toml]: I'm not sure this is related to this commit?

File: src/api/src/http/error.rs
Comment (sameo) [inline src/api/src/http/error.rs:103]: I think that belongs to the HTTP error handling improvement commit (9cccf7ab4b3dca86fa2172a26bb1da1c85bb299f)

File: docs/kbs_attestation_protocol.md
Comment (sameo) [reply docs/kbs_attestation_protocol.md:342]: > The former is usually
short-lived and is typically used for one-time, synchronous resources request. For longer KBS
sessions, it is more convenient to use the latter.

This may not be accurate at the moment, but I believe this is the intention: Being able to request resources with the token, regardless of the HTTP session status.

Comment (Xynnn007) [reply docs/kbs_attestation_protocol.md:342]: Thanks for the suggestion. It looks more comprehensible!

File: docs/kbs_attestation_protocol.md
Comment (Xynnn007) [reply docs/kbs_attestation_protocol.md:399]: Yes, thanks. I added some more details about the declaration.

File: src/api/Cargo.toml
Comment (Xynnn007) [reply src/api/Cargo.toml]: Oh, let me fix this. It is not used.

File: src/api/src/http/error.rs
Comment (Xynnn007) [reply src/api/src/http/error.rs:103]: Oh yes! Have rebased. Thanks!

**Comment by Xynnn007** [comment]
Hi @sameo @jialez0 , thanks for the suggestions and comments. I've updated the PR. PTAL

**Review by jialez0 — APPROVED** [review]
LGTM. Thanks @Xynnn007

**Review by sameo — APPROVED** [review]

**Comment by Xynnn007** [comment]
Hi @sameo sorry for not being clear. Could we revert this PR now? As it is during code freeze, I'm not sure whether this change would break other release components. Maybe we can re-merge this after the release?

**Comment by Xynnn007** [comment]
Emmm, let's retrigger the tdx CI and see if any new errors occur. If not, we do not need to revert. As revert is a dirty operaiton.

**Pull request merged**
